2021/day15/day15.java

```diff
@@ -0,0 +1,116 @@
+            this.risk = risk;
+        }
+
+        @Override
+        public int compareTo(Position other) {
+            return Integer.compare(this.risk, other.risk);
+        }
+    }
+
+    public static ArrayList<ArrayList<Integer>> readInput(String filename) {
+        ArrayList<ArrayList<Integer>> mapOfRisks = new ArrayList<>();
+        Scanner sc;
+        try {
+            sc = new Scanner(new File(filename));
+            while (sc.hasNextLine()) {
+                mapOfRisks.add(splitRow(sc.nextLine()));
+            }
+            sc.close();
+        } catch (Exception e) {
+            System.out.println("Error handling input " + e);
+        }
+        return mapOfRisks;
+    }
+
+    public static ArrayList<Integer> splitRow(String row) {
+        ArrayList<Integer> rowOfRisks = new ArrayList<>();
+        for (String s : new ArrayList<String>(Arrays.asList(row.split("")))) {
+            rowOfRisks.add(Integer.parseInt(s));
+        }
+        return rowOfRisks;
+    }
+
+    public static int getRisk(ArrayList<ArrayList<Integer>> mapOfRisks, int x, int y) {
+        int tmpRisk = mapOfRisks.get(y % mapOfRisks.size()).get(x % mapOfRisks.get(0).size())
+                + (x / mapOfRisks.get(0).size()) + (y / mapOfRisks.size());
+        return tmpRisk < 10 ? tmpRisk : ((tmpRisk % 10) + 1);
+    }
+
+    public static ArrayList<Point> getNeighbours(int x, int y, int width, int height) {
+        ArrayList<Point> neighbours = new ArrayList<>();
+        if (x > 0) {
+            neighbours.add(new Point(x - 1, y));
+        }
+        if (x < width - 1) {
+            neighbours.add(new Point(x + 1, y));
+        }
+        if (y > 0) {
+            neighbours.add(new Point(x, y - 1));
+        }
+        if (y < height - 1) {
+            neighbours.add(new Point(x, y + 1));
+        }
+        return neighbours;
+    }
+
+    public static int calculateRisk(boolean isPart2) {
+        ArrayList<ArrayList<Integer>> mapOfRisks = readInput("data.txt");
+        PriorityQueue<Position> positionQueue = new PriorityQueue<>();
+        int height = mapOfRisks.size();
+        int width = mapOfRisks.get(0).size();
+        if (isPart2) {
+            height *= 5;
+            width *= 5;
+        }
+        ArrayList<ArrayList<Boolean>> isVisited = new ArrayList<>();
+        for (int i = 0; i < height; i++) {
+            ArrayList<Boolean> line = new ArrayList<>(Collections.nCopies(width, false));
+            isVisited.add(line);
+        }
+        int x = 0;
+        int y = 0;
+        int risk = 0;
+        while (x != width - 1 || y != height - 1) {
+            for (Point point : getNeighbours(x, y, width, height)) {
+                Position position = new Position(point.x, point.y, risk + getRisk(mapOfRisks, point.x, point.y));
+                if (!isVisited.get(point.y).get(point.x)) {
+                    positionQueue.add(position);
+                }
+                isVisited.get(point.y).set(point.x, true);
```

Comment:
I find this slightly convoluted for a few reasons:

1. Why do we create the position before the check? Is it in any way necessary there?
2. Why is the last line outside of the check? Is that necessary?

What I would do is, before doing any logic of manipulating the queue, check if the neighbour is visited - if it is, `continue`, otherwise do all the logic at once, outside of any if.

Reply:
1. It is not. 🤣 
2. It is not. 🥇 

I like it more to check if "not visited" - and if it is in that state then do all the login inside if. That just seems more natural to me

Reply:
The problem with such an approach is that you then nest the `if`s inside one another in a more complicated scenario, making it less readable. Doing it this way (i.e. Stopping early if possible), you are guaranteed to have a more "linear-looking" code.

See https://youtu.be/n0Ak6xtVXno from 44:40 till around 46:30 for a comparison.